File: java/headFirstDesignPatterns/chap2/Weather-O-Rama/src/br/com/weather/WeatherData.java
package br.com.weather;

import java.util.ArrayList;

import br.com.headfirst.observer.Observer;
import br.com.headfirst.observer.Subject;

/**
 * Classe concreta que implementa Subject
 *
 */
public class WeatherData implements Subject {

    /**
     * Lista de Observers
     */
    private ArrayList<Observer> observers;
    
    /**
     * Temperatura
     */
    private float temperature;
    
    /**
     * Umidade
     */
    private float humidity;
    
    /**
     * Pressão
     */
    private float pressure;
    
    /**
     * Construtor padrão
     */
    public WeatherData() {
        this.observers = new ArrayList<>();
    }
    
    @Override
    public void registerObserver(Observer o) {
        this.observers.add(o);

    }

    @Override
    public void removeObserver(Observer o) {
        int i = this.observers.indexOf(o);
        
        if(i >= 0)
            this.observers.remove(i);
    }

    @Override
    public void notifyObservers() {
        for (Observer observer : this.observers) {
            observer.update(this.temperature, this.humidity, this.pressure);
        }
        
    }

    /**
     * Indica mudanças nas medidas, notificando Observers
     */
    public void measurementsChanged(){
        this.notifyObservers();
    }
    
    /**
     * Configura novas medidas
     * @param temperature Temperatura
     * @param humidity Umidade
     * @param pressure Pressão
     */
    public void setMeasurements(float temperature, float humidity, float pressure){
        this.temperature = temperature;
        this.humidity = humidity;
        this.pressure = pressure;
        this.measurementsChanged();
    }
}

File: c_cpp/headFirst_designPatterns/chap2/Weather-O-Rama/src/weather/WeatherData.cpp
/*
 * WeatherData.cpp
 *
 */

#include "WeatherData.hpp"

namespace weather{

WeatherData::WeatherData(float temp, float pres, float hum) {
    temperature = temp;
    pressure = pres;
    humidity = hum;
}

WeatherData::~WeatherData() {}

void WeatherData::addObserver(Object* observer){

    Observer* observerData = dynamic_cast<Observer*>(observer);
    if(observerData == 0)
        return;

    list<Observer*>::iterator it;
    for(it=observers.begin(); it!=observers.end();it++){
        if((*it) == observerData){
            return;
        }
    }
    observers.push_front(observerData);
}

void WeatherData::removeObserver(Object* observer){

    Observer* observerData = dynamic_cast<Observer*>(observer);
    if(observerData == 0)
        return;

    list<Observer*>::iterator it;
    for(it=observers.begin(); it!=observers.end();it++){
        if((*it) == observerData){
            it = observers.erase(it);
            return;
        }
    }
}

void WeatherData::notifyObservers(){
    list<Observer*>::iterator it;
    for(it=observers.begin(); it!=observers.end(); it++){
        (*it)->update(this);
    }
}

void WeatherData::measurementsChanged(){
    notifyObservers();
}

void WeatherData::setMeasurements(float temp, float hum, float pres){
    temperature = temp;
    humidity = hum;
    pressure = pres;
    measurementsChanged();
}

float WeatherData::getTemperature(){
    return temperature;
}

float WeatherData::getHumidity(){
    return humidity;
}

float WeatherData::getPressure(){
    return pressure;
}
}
